### PR TITLE
Support header/query encoding format

### LINKED ISSUE #1027: Update default collectionFormat and add encoding string to header and query parameters
Implementation issue for: https://github.com/Azure/cadl-azure/issues/1790

- Add optional parameters to query and header decorators for format
  - for query, format can be 'csv', 'multi', default should be multi
  - for header, only csv is allowed (and should be default)
  - target parameter/property must be a string array (or equivalent) type
- Add support in OpenAPI3 emitter for this additional data
  - for header, add `style: simple` to the parameter object (csv)
   - for query
     - add style: form for csv
     - add style: form and explode: true for multi
- Add support in cadl-autorest for this additional data (should emit collectionFormat: csv | multi in the schema)

### PR DESCRIPTION
**Pull request by leni-msft**
Relate with #1027 

TODOs in core:
- [x] Update decorator signatures. ~Reference: https://github.com/OAI/OpenAPI-Specification/blob/3.0.3/versions/3.0.3.md#style-values~
- [x] Update decorator implementations
- [x] Support encoding options in openapi3 emitter

### CODE REVIEW

File: packages/rest/lib/http-decorators.cadl [hunk 1]
```diff
@@ -2,45 +2,106 @@ namespace Cadl.Http;
...
+ */
+model HeaderOptions {
+  /**
+   * Name of the header when sent over http.
+   */
+  name?: string;
+  /**
+   * How the parameter value will be serialized
+   */
+  format?: "simple";
```
Comment (timotheeguerin) [inline packages/rest/lib/http-decorators.cadl:16]: I thought from the design we wanted to have this be `csv` not simple?

File: packages/rest/lib/http-decorators.cadl [hunk 2]
```diff
@@ -2,45 +2,106 @@ namespace Cadl.Http;
...
+  /**
+   * Whether parameter values of type `object` generate separate parameters for key-value pair of the map.
+   * For other types of parameters this property has no effect.
+   */
+  explode?: boolean;
```
Comment (timotheeguerin) [inline packages/rest/lib/http-decorators.cadl:21]: Isn't that `multi`?

File: packages/rest/lib/http-decorators.cadl (hunk 1 above)
Comment (leni-msft) [reply packages/rest/lib/http-decorators.cadl:16]: We could, but in that case, I think the decorator option will not be able to describe all possible formats in OpenAPI3.

For example, with `pipeDelimited` format: `blue|black|brown` for array, `R|100|G|200` for object.

Not sure if this has been considered?

Comment (leni-msft) [reply packages/rest/lib/http-decorators.cadl:16]: Confirmed with @markcowl , for now we'll only consider supporting limited formats in Cadl. I'll update the implementation accordingly.

File: common/changes/@cadl-lang/openapi3/fix-1027_2022-12-21-11-20.json [hunk 3]
```diff
@@ -0,0 +1,10 @@
+{
+  "changes": [
+    {
+      "packageName": "@cadl-lang/openapi3",
+      "comment": "Emit style & explode per encoding options",
+      "type": "none"
```
Comment (leni-msft) [inline common/changes/@cadl-lang/openapi3/fix-1027_2022-12-21-11-20.json:6]:
```suggestion
      "type": "patch"
```

**Review by markcowl — APPROVED** [review]

**Review by markcowl — CHANGES_REQUESTED** [review]
This looks great, but need to update https://github.com/Azure/cadl-azure/pull/2379  before this can be merged

**Comment by markcowl** [comment]
I have updated the cadl-azure PR here:  https://github.com/Azure/cadl-azure/pull/2425

**Comment by leni-msft** [comment]
Replaced with https://github.com/microsoft/cadl/pull/1527

**Pull request closed without merging**